README
Terminal-BASIC is a lightweight BASIC-like language interpreter for
microcontroller-based systems.  It was inspired by the TinyBASIC port
for popular 8-bit AVR uc boards.

This project aimed to create interpreter, partially compliant to the
standards ISO/IEC 6373:1984 and USSR/Russian ГОСТ 27787-88.	

1. Build from sources

Full source code is available at https://bitbucket.org/terminalbasicteam/terminal-basic
and https://sourceforge.net/p/terminal-basic/ repositories.
It depends on the libarduino, libarduinoext, libutft, libsdcard, tvoutex  and
libarduinoemulator libraries, which are being hosted at
    https://bitbucket.org/terminalbasicteam/libarduino.git
    https://bitbucket.org/terminalbasicte	am/libarduinoext.git
    https://bitbucket.org/terminalbasicteam/libutft.git
    https://bitbucket.org/terminalbasicteam/libsdcard.git
    https://bitbucket.org/terminalbasicteam/libtvoutex.git
    https://bitbucket.org/terminalbasicteam/libarduinoemulator.git
respectively (libs should be placed at the same directory, where terminal-basic
directory is placed i.e.
   .../REPO/-|
             |-terminal-basic from https://bitbucket.org/starling13/ucbasic/
             |-libarduino from https://bitbucket.org/starling13/libarduino/
             |-libarduinoext from https://bitbucket.org/starling13/libarduinoext/
             |-libutft from https://bitbucket.org/starling13/libutft/
             |-libsdcard from https://bitbucket.org/starling13/libsdcard/
             |-tvoutex from https://bitbucket.org/terminalbasicteam/libtvoutex.git
             |-libarduinoemulator from https://bitbucket.org/terminalbasicteam/libarduinoemulator.git )

The script ./make_sketch.sh makes an archive with arduino-ide sketch from source tree.

Use netbeans 8.2 IDE to work with sources. Building for PC (using
linux Arduino API emulation library) depends
on posixcpp library from https://bitbucket.org/starling13/posixcpp/
installed.

2. Installing and using

After building from sources or flashing precompiled hex file use appropriate input/output
devices, selected in config_arduino.hpp or config_linux.hpp before building.
By default terminal-basic use USART0 with default settings, 115200 baud rate.
Terminal program must:
    * support vt100 control codes
    * use no local echo and line editing.

Variants, using HX8357B based 480x320 TFT displays or composite video out
using UTFT library from http://www.rinkydinkelectronics.com/library.php?id=51
and TVoutEx library from https://starling13@bitbucket.org/starling13/tvoutex
respectively also works.

3. Terminal-BASIC language

A set of sources for test programs can be downloaded from
https://cloud.mail.ru/public/45SG/gGc2kFpBk

The language, which TerminalBASIC will support one day is:

Lexems:
 * OP_AND = "AND"       // 1
 * KW_ARRAYS = "ARRAYS" // 1
 * KW_BASE = "BASE"     // 2
 * COM_CHAIN = "CHAIN"  // 3
 * COM_CLS = "CLS"      // 4
 * KW_DATA = "DATA"     // 5
 * KW_DEF = "DEF"       // 6
 * COM_DELAY = "DELAY"  // 7
 * KW_DIM = "DIM"       // 8
 * COM_DUMP = "DUMP"    // 9
 * KW_END = "END"       // 10
 * KW_FALSE = "FALSE"   // 11
 * KW_FOR = "FOR"       // 12
 * KW_GOSUB = "GOSUB"   // 13
 * KW_GOTO = "GOTO"     // 14
 * KW_GO = "GO"         // 15
 * KW_IF = "IF"         // 16
 * KW_INPUT = "INPUT"   // 17
 * KW_LET = "LET"       // 18
 * COM_LIST = "LIST"    // 19
 * COM_LOAD = "LOAD"    // 20
 * COM_NEW = "NEW"      // 21
 * KW_NEXT = "NEXT"     // 22
 * OP_NOT = "NOT"
 * KW_ON = "ON"         // 23
 * KW_OPTION = "OPTION" // 24
 * OP_OR = "OR"
 * KW_PRINT = "PRINT"   // 25
 * KW_RANDOMIZE = "RANDOMIZE"
 * KW_READ = "READ"
 * KW_REM = "REM"
 * KW_RETURN = "RETURN"
 * COM_RUN = "RUN"
 * COM_SAVE = "SAVE"
 * KW_STEP = "STEP"
 * KW_STOP = "STOP"
 * KW_TAB = "TAB"
 * KW_THEN = "THEN"
 * KW_TO = "TO"
 * KW_TRUE = "TRUE"
 * KW_VARS = "VARS"
 * 
 * STAR = '*'
 * SLASH = '/'
 * PLUS = '+'
 * COLON = ':'
 * SEMI = ';'
 * LT = '<'
 * LTE = "<="
 * GT = '>'
 * GTE = ">="
 * EQUALS = '='
 * NE = "<>"
 * NEA = "><"
 * MINUS = '-'
 * POW = '^'
 * IDENT = [A-Z][A-Z0-9]*
 * C_INTEGER = [0-9]+
 * C_REAL = [0-9]+[.][0-9]*
 
Grammar:
 * TEXT = OPERATORS | C_INTEGER OPERATORS
 * OPERATORS = OPERATOR | OPERATOR COLON OPERATORS
 * OPERATOR =
 *	KW_DIM ARRAYS_LIST |
 *      KW_END |
 *	KW_FOR FOR_CONDS |
 *	KW_GOSUB EXPRESSION |
 *	KW_IF EXPRESSION IF_STATEMENT |
 *	KW_INPUT VAR_LIST |
 *	KW_LET IMPLICIT_ASSIGNMENT |
 *	KW_NEXT IDENT |
 *	KW_PRINT | KW_PRINT PRINT_LIST |
 *	KW_REM TEXT |
 *	KW_RETURN |
 *	KW_RANDOMIZE |
 *	GOTO_STATEMENT |
 *	COMMAND
 * COMMAND = COM_DUMP | COM_DUMP KW_VARS | COM_DUMP KW_ARRAYS
 *	COM_LIST | COM_NEW | COM_RUN | COM_SAVE | COM_LOAD
 * ASSIGNMENT = KW_LET IMPLICIT_ASSIGNMENT | IMPLICIT_ASSIGNMENT
 * IMPLICIT_ASSIGNMENT = VAR EQUALS EXPRESSION | VAR ARRAY EQUALS EXPRESSION
 * EXPRESSION = SIMPLE_EXPRESSION | SIMPLE_EXPRESSION REL SIMPLE_EXPRESSION
 * REL = LT | LTE | EQUALS | GT | GTE | NE | NEA
 * SIMPLE_EXPRESSION = TERM | TERM ADD TERM
 * ADD = PLUS MINUS KW_OR
 * TERM = FACTOR | FACTOR MUL FACTOR
 * MUL = STAR | SLASH | DIV | MOD | KW_AND
 * FACTOR = FINAL | FINAL POW FINAL
 * FINAL = C_INTEGER | C_REAL | C_STRING | VAR | VAR ARRAY |
 *	LPAREN EXPRESSION RPAREN | MINUS FINAL
 * VAR = REAL_IDENT | INTEGER_IDENT | STRING_IDENT
 * VAR_LIST = VAR | VAR VAR_LIST
 * PRINT_LIST = EXPRESSION | EXPRESSION COMMA PRINT_LIST
 * IF_STATEMENT = GOTO_STATEMEMNT | KW_THEN OPERATORS
 * GOTO_STATEMENT = KW_GOTO C_INTEGER
 * FOR_CONDS = IMPLICIT_ASSIGNMENT KW_TO EXPRESSION |
 *	IMPLICIT_ASSIGNMENT KW_TO EXPRESSION KW_STEP EXPRESSION
 * ARRAYS_LIST = VAR ARRAY | VAR ARRAY ARRAYS_LIST
 * ARRAY = LPAREN DIMENSIONS RPAREN
 * DIMENSIONS = C_INTEGER | C_INTEGER COMMA DIMENSIONS


ChangeLog
r.2.3-rc1 (12.04.2022)

  * Program text output changed: extra spaces were removed
  * Fix AVR8 external memory support
  * Fix support of the SD-card on Arduino platforms
  * Fix Arduino library-based implementation of the NOTONE command
  * Fix control of the active terminal on multiple I/O devices systems
  * Fix regression in AUTORUN.BAS support

r2.3-b2 (22.11.2021)

  * Add VAL function
  * Fix DATA ... READ with non-quoted strings
  * Fix bug in ABS function with long REAL values

r2.3-b1 (06.04.2021)

  // Platforms
  * Arduino: add support of PS/2 keyboard as input device
  * Add initial support of ODROID-GO device  with optional QWERTY keyboard

  // Bug fixes:
  * Fix bug in Lexer, which causes fake string overflow errors
  * Fix bug in matrix assignment in the program line with colon separator
  * Fix buffer overflow in printing of double precision constants

r2.3-a2 (26.02.2021) Release, dedicated to 4 years of TB and 4k downloads from sourceforge.net

  * Output to SPI TFT screens via Adafruit_ILI9341 and TFT_eSPI libs
  * ESP32: add DTONE/NOTONE implementation
  * Working MS Windows port

r2.3-a1 (18.10.2020)

  * Introduction of HAL layer, simplifying the process of porting to different platforms
  * Add double precision math functions

r2.2.2 (06.04.2021)

  * Fix bug in Lexer, which causes fake string overflow errors (backport from
    2.3 branch)
  * Fix bug in matrix assignment in the program line with colon separator
     (backport from 2.3 branch)
  * Fix buffer overflow in printing of double precision constants (backport from
    2.3 branch)

r2.2.1 (08.03.2020) As always dedicated to International women day.
    Thank you, our sweets for frendship, love and children

  * Fix bug with printing first command in module with fast module call mode
  * Fix wrong order of GFX commands with explicit color commands enabled

r2.2 (23.02.2020) Release, dedicated tpo the Day of the Soviet army and NAVY

  * Add support of the string arrays
  * TVout output uses cp866 code page font by default

r2.2-rc1 (18.02.2020)

  * Fix bug with multidimentional arrays with one of the dimensions >= 255
  * Implement optional compilation of the GFX commands with explicit color setting

r2.2-beta3 (16.02.2020)

  * Add parameters to DIRECTORY command (optional 1st and last file numbers
    to output)
  * Fix bugs with alignmnent. ESP8266 platform has reasonable suppot.

r2.2-beta2 (12.02.2020)

  * ESP8266 and ESP32 supports SAVE/LOAD commands in SPIFFS
  * improved optional alignment of values and function parameters

r2.2-beta1 (7.02.2020) Release, dedicated to my sisters Birthday

  * Fix error with fast modules calling on some 32-bit platforms

r2.2-alpha3 (5.02.2020)

  * Add optional feature of aligned variable and array frames (step
    towards optional removal of unaligned memory access)
  * Basic support of esp8266
  * Add support of long real datatype in matrix operations

r2.2-alpha2 (24.01.2020)

  * Fix multiple regressions

r2.2-alpha1 (22.01.2020)

  * Speedup of the commands from modules calling (ArduinoIO, GFX ...), using translation
      of the calling construct into buinary optimized format

r2.1.1 (15.10.2019)

  * Fix error in configuratyion parameters on tiny systems
  * Basic support of ESP32

r2.1 (23.09.2019)

  * Add rudimentary support of french keywords
  * Fix RETURN, called from FOR LOOP
  * Add ELLIPSE and ELLIPSEC commands to gfx module

r2.1-beta1 (29.05.2019)

  * Add support for implicit arrays with indexes from 0 to 10
  * Add command FSEEK to SD module
  * Fix bug with function calls using long real parameters
  * Fix bug with confirmation of SCRATCH/HEADER

r2.1-alpha2 (22.04.2019)

  * Fix regression in functions call when using long reals

r2.1-alpha1 (22.04.2019)

  * Fix FOR loop limit and increment evaluation order (ANSI standad 11.2, 11.4)
  * Change logical variables and arrays ending from "!" to "@"
  * Change long integer variables and arrays ending from "%%" to "%!"
  * Change GET$ function to more common INKEY$
  * Add partial support of long reals (variable/array name ends with "!")
  * Add functions MID$/SEG$

r2.0.3 (9.05.2019 Release, dedicated to the USSR Victory Day

  * Fix error in confirmation of the file operations
  * Fix regression in SD functions while reading the files with "CR LF" line endings
  * Output reals with 6 significant decimal digits

r2.0.2 (24.04.2019)

  * Fix bugs in building with ArduinoIDE, using TVOUT output

r2.0.1 (21.04.2019)

  * Fix bug with parsing real constants with E/e and w/o decimal point
  * Fix bug wirh saving program by DSAVE with REM lines

r2.0 (11.04.2019)

  * Reduce unnecessary spaces, generated in source file with DSAVE
  * Fix undefined behaviour when load from file with long lines

r2.0-rc1 (24.03.2019)

  * Add binary file operations to SD module
  * Add HEX$ function
  * Update libsdcard (No more dependency on arduino SD lib)

r2.0-beta-3 (15.03.2019)

  * Fix 2.0 regression in parsing reals with E and e exponent chars

r2.0-beta-2 (9.03.2019)

  * Fix lexer bug with input errors handling

r2.0-beta-1 (8.03.2019) Release, dedicated to the international women day

  * Add autostart of the AUTORUN.BAS program if SD support is enabled
  * Fix SD module DSAVE regression
  * Fix incorrect matrix multiplication restriction
  * Fix scanning of identifiers with char codes greater then 127

r2.0-alpha-3 (23.02.2019) Release, dedicated to the Defender of the Fatherlad Day

  * Fix commands and functions parameters passing
  * Fix a conflict of GET$() FUNCTION and DELAY command
  * Fix some lexer regressions
  * Add more russian keywords from BASIC-6 system for BESM-6 computer

r2.0-alpha-2 (19.02.2019)

  * Fix some lexer-related regressions

r2.0-alpha-1 (17.02.2019)

  * New lexical analyzer, written in pure C. Ability to localize keywords.
    Currently English and Russian keywords  available.

r1.2 (16.02.2019) Release, dedicated to 43 anniversary of the first built KAMAZ truck


terminal-basic/basic.hpp
/*
 * This file is part of Terminal-BASIC: a lightweight BASIC-like language
 * interpreter.
 * 
 * Copyright (C) 2016-2018 Andrey V. Skvortsov <starling13@mail.ru>
 * Copyright (C) 2019-2021 Terminal-BASIC team
 *     <https://github.com/terminal-basic-team>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef BASIC_COMMON_HPP
#define BASIC_COMMON_HPP

#include <stdio.h>
#include <inttypes.h>

#include <Arduino.h>
#include "HAL.h"

#include "tools.h"
#include "types.hpp"
#include "HALProxyStream.hpp"

#ifdef ARDUINO
#include "config_arduino.hpp"
#elif (defined __linux__) || \
      (defined __MINGW32__)
#include "config_linux.hpp"
#endif

#include "basic.h"

#if USE_EXTEEPROM
	#define USE_WIRE 1
#else
	#define USE_WIRE 0
#endif

#if S_INPUT == HAL_I
	#define SERIAL_PORT_I halproxy1
#endif

#ifdef SERIAL_PORT_I
	#define INPUT_STREAM SERIAL_PORT_I
#endif

#if S_OUTPUT == HAL_O
	#define SERIAL_PORT_O halproxy1
#endif // S_OUTPUT

#ifdef SERIAL_PORT_O
	#define OUTPUT_PRINT SERIAL_PORT_O
#endif

#ifdef true
#undef true
#endif

#ifdef false
#undef false
#endif

/**
 * @brief Simple BASIC language interpreter package
 */
namespace BASIC
{
typedef ::pointer_t Pointer;
// integer type
typedef ::integer_t Integer;
constexpr const Integer MaxInteger = MAX_INTEGER;
#if USE_LONGINT
// long integer type
typedef ::long_integer_t LongInteger;
constexpr const LongInteger MaxLongInteger = MAX_LONG_INTEGER;
#endif // USE_LONGINT
typedef ::INT INT;
// floating point type
#if USE_REALS
using Real = real_t;
#if USE_LONG_REALS
using LongReal = long_real_t;
#endif
#endif // USE_REALS

#if USE_PACKED_STRUCT
#define PACKED EXT_PACKED
#else
#define PACKED
#endif

// Number of characters in command/function identifier
const uint8_t IDSIZE = 8;

enum class ProgMemStrings : uint8_t
{
	S_AT = 0,
	S_BYTES,
	S_STATIC,
	S_DYNAMIC,
	S_ERROR,
	S_SEMANTIC,
	S_SYNTAX,
	S_READY,
	AVAILABLE,
	TERMINAL,
	S_TERMINAL_BASIC,
	S_VERSION,
	S_TEXT,
	S_OF,
#if USE_DUMP
	S_VARS,
	S_ARRAYS,
	S_STACK,
#endif // USE_DUMP
	S_REALLY,
	S_END,
#if USE_TEXTATTRIBUTES
        VT100_ESCSEQ,
	VT100_CLS,
	VT100_NOATTR,
	VT100_BRIGHT,
	VT100_UNDERSCORE,
	VT100_REVERSE,
	VT100_LINEHOME,
#if SET_PRINTZNES
	VT100_CLEARZONES,
	VT100_SETZONE,
#endif // SET_PRINTZNES
#if USE_COLORATTRIBUTES
	VT100_RED,
	VT100_GREEN,
	VT100_YELLOW,
	VT100_BLUE,
	VT100_MAGENTA,
	VT100_CYAN,
	VT100_WHITE,
#endif // USE_COLORATTRIBUTES
#endif // USE_TEXTATTRIBUTES
	NUM_STRINGS
};

// Static text strings
extern PGM_P progmemString(ProgMemStrings);

/**
 * @brief lexical tokens
 */
enum class Token : uint8_t
{
	NOTOKENS = BASIC_TOKEN_NOTOKEN,    // 0
	OP_AND = BASIC_TOKEN_OP_AND,       // 1
#if USE_DUMP
	KW_ARRAYS = BASIC_TOKEN_KW_ARRAYS, // 2
#endif
//	KW_BASE,       // 3
#if USE_SAVE_LOAD
	COM_CHAIN = BASIC_TOKEN_COM_CHAIN, // 4
#endif
#if USE_TEXTATTRIBUTES
	COM_CLS = BASIC_TOKEN_COM_CLS,     // 5
#endif
#if USESTOPCONT
	COM_CONT = BASIC_TOKEN_COM_CONT,   // 6
#endif
#if USE_MATRIX
	KW_CON = BASIC_TOKEN_KW_CON,       // 7
#endif
#if USE_DATA
	KW_DATA = BASIC_TOKEN_KW_DATA,     // 8
#endif
#if USE_DEFFN
	KW_DEF = BASIC_TOKEN_KW_DEF,       // 9
#endif
#if USE_DELAY
	COM_DELAY = BASIC_TOKEN_COM_DELAY,// 10
#endif
#if USE_MATRIX
	KW_DET = BASIC_TOKEN_KW_DET,      // 11
#endif
	KW_DIM = BASIC_TOKEN_KW_DIM,      // 12
#if USE_DIV_KW
	KW_DIV = BASIC_TOKEN_KW_DIV,      // 13
#endif
#if USE_DOLOOP
	KW_DO = BASIC_TOKEN_KW_DO,        // 14
#endif
#if USE_DUMP
	COM_DUMP = BASIC_TOKEN_COM_DUMP,  // 15
#endif
	KW_END = BASIC_TOKEN_KW_END,      // 16
	KW_FALSE = BASIC_TOKEN_KW_FALSE,  // 17
#if USE_DEFFN
	KW_FN = BASIC_TOKEN_KW_FN,        // 18
#endif
	KW_FOR = BASIC_TOKEN_KW_FOR,      // 19
	KW_GOSUB = BASIC_TOKEN_KW_GOSUB,  // 20
	KW_GOTO = BASIC_TOKEN_KW_GOTO,    // 21
#if CONF_SEPARATE_GO_TO
	KW_GO = BASIC_TOKEN_KW_GO,        // 22
#endif
#if USE_MATRIX
	KW_IDN = BASIC_TOKEN_KW_IDN,      // 23
#endif
	KW_IF = BASIC_TOKEN_KW_IF,        // 24
	KW_INPUT = BASIC_TOKEN_KW_INPUT,  // 25
#if USE_MATRIX
	KW_INV = BASIC_TOKEN_KW_INV,      // 26
#endif
	KW_LET = BASIC_TOKEN_KW_LET,      // 27
	COM_LIST = BASIC_TOKEN_COM_LIST,  // 28
#if USE_SAVE_LOAD
	COM_LOAD = BASIC_TOKEN_COM_LOAD,  // 29
#endif
#if USE_TEXTATTRIBUTES
	COM_LOCATE = BASIC_TOKEN_COM_LOCATE,// 30
#endif
#if USE_DOLOOP
	KW_LOOP = BASIC_TOKEN_KW_LOOP,      // 31
#endif
#if USE_MATRIX
	KW_MAT = BASIC_TOKEN_KW_MAT,        // 32
#endif
#if USE_INTEGER_DIV
	KW_MOD = BASIC_TOKEN_KW_MOD,        // 33
#endif
	COM_NEW = BASIC_TOKEN_COM_NEW,      // 34
	KW_NEXT = BASIC_TOKEN_KW_NEXT,      // 35
	OP_NOT = BASIC_TOKEN_OP_NOT,        // 36
#if CONF_USE_ON_GOTO
	KW_ON = BASIC_TOKEN_KW_ON,          // 37
#endif
//	KW_OPTION,     // 38
	OP_OR = BASIC_TOKEN_OP_OR,
#if USE_PEEK_POKE
	KW_POKE = BASIC_TOKEN_KW_POKE,
#endif
	KW_PRINT = BASIC_TOKEN_KW_PRINT,    // 40
#if USE_RANDOM
	KW_RANDOMIZE = BASIC_TOKEN_KW_RANDOMIZE,// 41
#endif
#if USE_DATA
	KW_READ = BASIC_TOKEN_KW_READ,      // 42
#endif
	KW_REM = BASIC_TOKEN_KW_REM,        // 43
#if USE_DATA
	KW_RESTORE = BASIC_TOKEN_KW_RESTORE,// 44
#endif
	KW_RETURN = BASIC_TOKEN_KW_RETURN,  // 45
	COM_RUN = BASIC_TOKEN_COM_RUN,      // 46
#if USE_SAVE_LOAD
	COM_SAVE = BASIC_TOKEN_COM_SAVE,    // 47
#endif
#if CONF_USE_SPC_PRINT_COM
	KW_SPC = BASIC_TOKEN_KW_SPC,        // 48
#endif
	KW_STEP = BASIC_TOKEN_KW_STEP,      // 49
#if USESTOPCONT
	KW_STOP = BASIC_TOKEN_KW_STOP,      // 50
#endif
#if USE_TEXTATTRIBUTES
	KW_TAB = BASIC_TOKEN_KW_TAB,        // 51
#endif
	KW_THEN = BASIC_TOKEN_KW_THEN,      // 52
	KW_TO = BASIC_TOKEN_KW_TO,         // 53
#if USE_MATRIX
	KW_TRN = BASIC_TOKEN_KW_TRN,        // 54
#endif
	KW_TRUE = BASIC_TOKEN_KW_TRUE,      // 55
#if USE_DUMP
	KW_VARS = BASIC_TOKEN_KW_VARS,      // 56
#endif
	OP_XOR = BASIC_TOKEN_OP_XOR,        // 57
#if USE_MATRIX
	KW_ZER = BASIC_TOKEN_KW_ZER,        // 58
#endif
	// *
	STAR = BASIC_TOKEN_STAR,          // 59
	// /
	SLASH = BASIC_TOKEN_SLASH,         // 50
#if USE_REALS && USE_INTEGER_DIV
	BACK_SLASH = BASIC_TOKEN_BACK_SLASH,    // 61
#endif
	// +
	PLUS = BASIC_TOKEN_PLUS,          // 62
	// -
	MINUS = BASIC_TOKEN_MINUS,         // 63
	// =
	EQUALS = BASIC_TOKEN_EQUALS,        // 64
	// :
	COLON = BASIC_TOKEN_COLON,         // 65
	// ;
	SEMI = BASIC_TOKEN_SEMI,          // 66
	// <
	LT = BASIC_TOKEN_LT,            // 67
	// >
	GT = BASIC_TOKEN_GT,            // 68
	// <=
	LTE = BASIC_TOKEN_LTE,           // 69
	// >=
	GTE = BASIC_TOKEN_GTE,           // 70
	// <>
	NE = BASIC_TOKEN_NE,            // 71
#if CONF_USE_ALTERNATIVE_NE
	//  ><
	NEA = BASIC_TOKEN_NEA,           // 72
#endif
	// ,
	COMMA = BASIC_TOKEN_COMMA,         // 73
	// ^
	POW = BASIC_TOKEN_POW,           // 74
	// (
	LPAREN = BASIC_TOKEN_LPAREN,        // 75
	// )
	RPAREN = BASIC_TOKEN_RPAREN,        // 76

	INTEGER_IDENT = BASIC_TOKEN_INTEGER_IDENT, // 77
	REAL_IDENT = BASIC_TOKEN_REAL_IDENT,    // 78
#if USE_LONG_REALS
	LONG_REAL_IDENT = BASIC_TOKEN_LONG_REAL_IDENT,
#endif
#if USE_LONGINT        // 79
	LONGINT_IDENT = BASIC_TOKEN_LONGINT_IDENT, // 80
#endif
	STRING_IDENT = BASIC_TOKEN_STRING_IDENT,  // 81
	BOOL_IDENT = BASIC_TOKEN_BOOL_IDENT,    // 82

	C_INTEGER = BASIC_TOKEN_C_INTEGER,     // 83
#if USE_LONGINT
	C_LONG_INTEGER = BASIC_TOKEN_C_LONG_INTEGER, // 83
#endif
	C_REAL = BASIC_TOKEN_C_REAL,           // 84
#if USE_LONG_REALS
	C_LONG_REAL = BASIC_TOKEN_C_LONG_REAL,
#endif
	C_BOOLEAN = BASIC_TOKEN_C_BOOLEAN,     // 85
	C_STRING = BASIC_TOKEN_C_STRING,       // 86
#if FAST_MODULE_CALL
	COMMAND = BASIC_TOKEN_COMMAND,         // 87
#endif
	NUM_TOKENS = BASIC_TOKEN_NUM_TOKENS    // 88
};

template <uint8_t s>
inline typename type_factory<s>::unsigned_type readValueOfSize(const uint8_t*)
{
	return 0;
}

template <>
inline uint64_t readValueOfSize<8>(const uint8_t* str)
{
	uint64_t result;
	readU64(&result, str);
	return result;
}

template <>
inline uint32_t readValueOfSize<4>(const uint8_t* str)
{
	uint32_t result;
	readU32(&result, str);
	return result;
}

template <>
inline uint16_t readValueOfSize<2>(const uint8_t* str)
{
	uint16_t result;
	readU16(&result, str);
	return result;
}

template <typename T>
inline T readValue(const uint8_t* str)
{
	return readValueOfSize<sizeof(T)>(str);
}

template <uint8_t s>
inline void writeValueOfSize(typename type_factory<s>::unsigned_type v, uint8_t* str)
{
}

template <>
inline void writeValueOfSize<8>(uint64_t val, uint8_t* str)
{
	writeU64(val, str);
}

template <>
inline void writeValueOfSize<4>(uint32_t val, uint8_t* str)
{
	writeU32(val, str);
}

template <>
inline void writeValueOfSize<2>(uint16_t val, uint8_t* str)
{
	writeU16(val, str);
}

template <typename T>
inline void writeValue(T v, uint8_t* str)
{
	writeValueOfSize<sizeof(T)>(v, str);
}

#if CONF_USE_ALIGN
#define READ_VALUE(V) readValue<decltype(V)>(reinterpret_cast<const uint8_t*>(&(V)))
#define WRITE_VALUE(D, V) writeValue(V, reinterpret_cast<uint8_t*>(&D))
#else
#define READ_VALUE(V) (V)
#define WRITE_VALUE(D, V) D = (V)
#endif // CONF_USE_ALIGN

} // namespace BASIC

#endif // BASIC_COMMON_HPP
